### ISSUE TITLE
Wrong allocation of intent(out) arguments

### ISSUE THREAD

**Issue by yguclu** [pyccel/pyccel#174]
When translating a Python function that returns an array, Pyccel generates a Fortran subroutine with an `intent(out)` array variable and an `allocate` statement for it. But such array is not declared as allocatable, and hence the Fortran code cannot be compiled. This is related to issue #154.

**Comment by yguclu** [comment]
This happens when the output array is created inside the function with a Numpy instruction like `empty`, `zeros` or `ones`. For example, the Python function
```
#$ header function twos( int ) results( int[:] )
def twos( n ):
    from numpy import empty
    a = empty( n, dtype=int )
    a[:] = 2
    return a
```
is translated to the Fortran code
```
subroutine twos(n, a)

  implicit none
  integer(kind=4), intent(out)  :: a (0:)
  integer(kind=4), intent(in)  :: n


  allocate(a(0:n - 1))
  a(:) = 2
  return
end subroutine
```
Which cannot compile because `a` is not an allocatable variable. If we want the Fortran subroutine to take care of **memory allocation**, then we should declare `a` as `allocatable`. Instead, if we want to call this Fortran subroutine from Python (e.g. by using **f2py** through **epyccel**), then the allocation will be performed in the Python middle layer generated by **f2py**, and the `allocate` statement should be removed:
```
subroutine twos(n, a)
  integer(kind=4), intent(in ) :: n
  integer(kind=4), intent(out) :: a(0:n)

  a(:) = 2
end subroutine
```
@saidctb @ratnania I think this is a delicate issue that may require some in-depth discussion.

**Comment by saidctb** [comment]
@yguclu  from what @ratnania  told me when we were discussing the rules,  that we won't allow the allocation inside the function and return it after that all the returned array must be allocated outside of the function.

**Comment by yguclu** [comment]
Yes, f2py will take care of allocation for us.

**Comment by yguclu** [comment]
@EmilyBourne Can you please check the status of this old issue?

**Comment by EmilyBourne** [comment]
@yguclu 
The example code still fails and produces the following output:
```
Array return arguments are currently not supported

ERROR at annotation (semantic) stage
Traceback (most recent call last):
  File "/home/emily/.local/bin/pyccel", line 11, in <module>
    load_entry_point('pyccel==0.9.7', 'console_scripts', 'pyccel')()
  File "/home/emily/.local/lib/python3.6/site-packages/pyccel/commands/console.py", line 225, in pyccel
    folder        = args.output)
  File "/home/emily/.local/lib/python3.6/site-packages/pyccel/codegen/pipeline.py", line 116, in execute_pyccel
    ast = parser.annotate(**settings)
  File "/home/emily/.local/lib/python3.6/site-packages/pyccel/parser/parser.py", line 104, in annotate
    **settings)
  File "/home/emily/.local/lib/python3.6/site-packages/pyccel/parser/semantic.py", line 190, in __init__
    self.annotate()
  File "/home/emily/.local/lib/python3.6/site-packages/pyccel/parser/semantic.py", line 226, in annotate
    ast = self._visit(ast, **settings)
  File "/home/emily/.local/lib/python3.6/site-packages/pyccel/parser/semantic.py", line 1134, in _visit
    return getattr(self, annotation_method)(expr, **settings)
  File "/home/emily/.local/lib/python3.6/site-packages/pyccel/parser/semantic.py", line 1148, in _visit_Tuple
    ls = [self._visit(i, **settings) for i in expr]
  File "/home/emily/.local/lib/python3.6/site-packages/pyccel/parser/semantic.py", line 1148, in <listcomp>
    ls = [self._visit(i, **settings) for i in expr]
  File "/home/emily/.local/lib/python3.6/site-packages/pyccel/parser/semantic.py", line 1134, in _visit
    return getattr(self, annotation_method)(expr, **settings)
  File "/home/emily/.local/lib/python3.6/site-packages/pyccel/parser/semantic.py", line 2739, in _visit_FunctionDef
    raise AstFunctionResultError(r)
pyccel.ast.core.AstFunctionResultError: Found allocatable result(s) that is/are not inout [a]
```

If we plan to allow returning array arguments then this issue is still valid. If we don't then the error message should be handled by the error module instead of crashing with a traceback

**Comment by yguclu** [comment]
@EmilyBourne: Thanks. I am adding this issue to the "Numpy support" milestone

**Comment by EmilyBourne** [comment]
@yguclu The example code now generates the following error message:
```

ERROR at annotation (semantic) stage
pyccel: [semantic] 
 |fatal: [2,1-0]| Array return arguments are currently not supported (a)

```
Do we consider this fixed by PR #350?

**Comment by yguclu** [comment]
Yes, given that we don't allow array return arguments at the moment, I consider this issue fixed.
